### ISSUE TITLE
Vulnerable packages are not being detected in Arch Linux

### ISSUE THREAD

**Issue by mcarmona99** [wazuh/wazuh#13775]
|Wazuh version|Component|Platform|
|---|---|---|
| 4.3.4-rc1 | Vulnerability detector | Arch Linux Agent |

After the E2E UX tests done for the RC1 of Wazuh version 4.3.4 in https://github.com/wazuh/wazuh/issues/13688, an issue related to vulnerability detection in an Arch Linux endpoint has been found. If we try to install a vulnerable package, the vulnerability is not detected by the Vulnerability detector.

The environment used to test the vulnerability detector module was: 1 Wazuh manager (Amazon Linux 2) with 1 Wazuh agent connected (Arch Linux).

After the configuration, VD detected 7 vulnerabilities:

![image](https://user-images.githubusercontent.com/57481331/172398133-fc947030-5e28-4531-bc56-372c3e9bbc23.png)

I installed a vulnerable package (one of the listed in https://security.archlinux.org/). The package chosen was `openssl`. There is an online repository that provides old packages for the Arch Linux OS: https://archive.archlinux.org/packages/. I installed **OpenSSL 1.1.1.n-1** which is affected by [CVE-2022-1292](https://security.archlinux.org/CVE-2022-1292). Upgrading to **OpenSSL 1.1.1.o-1** patches the vulnerability.

Package installation:

```
# curl -sO https://archive.archlinux.org/packages/o/openssl/openssl-1.1.1.n-1-x86_64.pkg.tar.zst
# pacman -U openssl-1.1.1.n-1-x86_64.pkg.tar.zst 
loading packages...
warning: downgrading package openssl (1.1.1.o-1 => 1.1.1.n-1)
resolving dependencies...
looking for conflicting packages...

Packages (1) openssl-1.1.1.n-1

Total Installed Size:  7.37 MiB
Net Upgrade Size:      0.00 MiB

:: Proceed with installation? [Y/n] y
...
```

After the installation:

```
# openssl version
OpenSSL 1.1.1n  15 Mar 2022
```

We can also see the installed package in the agent's inventory data:

![image](https://user-images.githubusercontent.com/57481331/172406870-45275f72-87b2-421b-be7e-2c452d772284.png)

Vulnerability: NO vulnerability detected

Alert: NO alert generated because no vulnerability was detected

The vulnerability detector did not detected this specific vulnerability.

We can see in https://security.archlinux.org/:

![image](https://user-images.githubusercontent.com/57481331/172406334-8d2270cf-249a-411e-9cf6-313b4373cb7d.png)

And the NVD database also includes this vulnerability: https://nvd.nist.gov/vuln/detail/CVE-2022-1292

![image](https://user-images.githubusercontent.com/57481331/172406637-b7280351-8481-4d9d-bd15-8e4ddd3e7a36.png)

V. Detector properly configured:

```xml
<provider name="arch">
      <enabled>yes</enabled>
      <update_interval>10m</update_interval>
    </provider>

...

<provider name="nvd">
      <enabled>yes</enabled>
      <update_from_year>2010</update_from_year>
      <update_interval>10m</update_interval>
    </provider>
```

I also did the same test but with a vulnerable version of `containerd`. The vulnerability was not detected neither.

**Comment by MarcelKemp** [comment]
## Research details

After researching vulnerability scanning for an Arch Linux agent, I was able to obtain the following information:

> Keep in mind that the feeds and versions are updated every day and these cases will remain deprecated, but with these cases the different situations found are explained.

- #### According to the package reported in the issue description (OpenSSL version `1.1.1.1.n-1`), for CVE-2022-1292:

  In this case, as indicated in the Arch Linux feed, the version is among those affected, and also has a patch that fixes this vulnerability.
  The problem appears due to the vulnerability's status: `Unknown`.
  This causes the Vulnerability Detector, not being in a state of `Vulnerable` or `Fixed`, to discard the vulnerability.
  This `Unknown` status is only found in new vulnerabilities that are added to the feed or in those vulnerabilities that are under reanalysis, so in order to avoid false positive alerts, scanning is avoided until this status is modified, when the vulnerability will be reported correctly.

  - OVAL (in `Unknown` state): https://security.archlinux.org/CVE-2022-1292
  - NVD (`Undergoing reanalysis`): https://nvd.nist.gov/vuln/detail/CVE-2022-1292

- #### OVAL vulnerabilities with `Vulnerable` status and with a fixed version:

  Cases have been found in OVAL, where packages with the status `Vulnerable` appear next to a higher version of the package that fixes the vulnerability. The problem here is that it seems that the vulnerability should be fixed, but after performing several tests, if OVAL has not changed the status to `Fixed`, then the higher version is not available in the official repository.

  In these cases, the vulnerability detector listens to OVAL, so that if the `Vulnerable` status appears, then we only check that the package is higher or equal to the affected version, as the fix is not yet officially available.

  Example:

  - The following link shows all vulnerabilities with status `Vulnerable` from OVAL: https://security.archlinux.org/issues

  - Below, the _Vulnerability Detector_ results for several of these packages (`subversion`, `thunderbird` and `containerd`):

  ```
  DEBUG: (5468): The 'subversion' package (1.14.1-6) from agent '001' is vulnerable to 'CVE-2022-24070'. Condition: 'Package greater than or equal 1.14.1-6'
  DEBUG: (5468): The 'thunderbird' package (91.9.1-1) from agent '001' is vulnerable to 'CVE-2022-1834'. Condition: 'Package greater than or equal 91.9.1-1'
  DEBUG: (5468): The 'thunderbird' package (91.9.1-1) from agent '001' is vulnerable to 'CVE-2022-31736'. Condition: 'Package greater than or equal 91.9.1-1'
  DEBUG: (5468): The 'thunderbird' package (91.9.1-1) from agent '001' is vulnerable to 'CVE-2022-31737'. Condition: 'Package greater than or equal 91.9.1-1'
  DEBUG: (5468): The 'thunderbird' package (91.9.1-1) from agent '001' is vulnerable to 'CVE-2022-31738'. Condition: 'Package greater than or equal 91.9.1-1'
  DEBUG: (5468): The 'subversion' package (1.14.1-6) from agent '001' is vulnerable to 'CVE-2021-28544'. Condition: 'Package greater than or equal 1.14.1-6'
  DEBUG: (5468): The 'thunderbird' package (91.9.1-1) from agent '001' is vulnerable to 'CVE-2022-31741'. Condition: 'Package greater than or equal 91.9.1-1'
  DEBUG: (5468): The 'thunderbird' package (91.9.1-1) from agent '001' is vulnerable to 'CVE-2022-31742'. Condition: 'Package greater than or equal 91.9.1-1'
  DEBUG: (5468): The 'thunderbird' package (91.9.1-1) from agent '001' is vulnerable to 'CVE-2022-31747'. Condition: 'Package greater than or equal 91.9.1-1'
  DEBUG: (5468): The 'containerd' package (1.6.5-1) from agent '001' is vulnerable to 'CVE-2022-31030'. Condition: 'Package greater than or equal 1.6.5-1'
  ```
  And as can be seen, the officially supported package for e.g. `containerd` in the Arch Linux repository, still remains the `1.6.5-1` version (even though it shows that the fixed version is `1.6.6-1`): https://security.archlinux.org/package/containerd

  So, when they upload the new version to the repository, that vulnerability will move from `Vulnerable` to `Fixed` status in the OVAL.

- #### Problems with some versioning of NVD packages according to the official OVAL:

  Finally, a small problem has been observed in the versioning comparison between the original package and the CPE entry in the NVD. It seems that the NVD does not have the same versioning scheme for some packages, which causes that when comparing the versioning with the NVD, some packages are matched and others are discarded when they should not be.

  In the case of Arch Linux, re is no problem for _pacman_ packages as the OVAL is fully trusted, so the NVD vulnerabilities would only reconfirm that they are indeed vulnerable, but in no case discards or adds any vulnerabilities, so we avoid false positives and false negatives from the NVD.

  Examples of Vulnerability Detector traces for some _pacman_ packages: 

  - **`openssl` package** (wrong version scheme by NVD `1.1.1x` instead of `1.1.1.x-1`):
  ```
  DEBUG: (5460): Package 'openssl' not vulnerable to 'CVE-2021-3450'. Version (1.1.1.o-1) not 'less than' '1.1.1.k-1' (feed 'OVAL').
  DEBUG: (5460): Package 'openssl' not vulnerable to 'CVE-2021-23840'. Version (1.1.1.o-1) not 'less than' '1.1.1.j-1' (feed 'OVAL').
  ...
  DEBUG: (5460): Package 'openssl' not vulnerable to 'CVE-2021-3450'. Version (1.1.1.o-1) not 'greater than or equal' '1.1.1h' (feed 'NVD').
  DEBUG: (5458): Package 'openssl' inserted into the vulnerability 'CVE-2021-23840'. Version (1.1.1.o-1) 'less than' '1.1.1j' (feed 'NVD').
  DEBUG: (5458): Package 'openssl' inserted into the vulnerability 'CVE-2022-1292'. Version (1.1.1.o-1) 'less than' '1.1.1o' (feed 'NVD').
  ...
  DEBUG: (5462): Package 'openssl' not vulnerable to 'CVE-2021-23840' since it is not affected (feed 'OVAL').
  DEBUG: (5462): Package 'openssl' not vulnerable to 'CVE-2022-1292' since it is not affected (feed 'OVAL').
  ```
  1. [CVE-2021-3450](https://security.archlinux.org/CVE-2021-3450): By OVAL not vulnerable because version `1.1.1.o-1` not 'less than' `1.1.1.k-1`. Then the NVD says it is not vulnerable either - `1.1.1.o-1` not 'greater than or equal to' **`1.1.1h`**. Although, as it has been discarded by OVAL, then it is not considered vulnerable.
  2. [CVE-2021-23840](https://security.archlinux.org/CVE-2021-23840): By OVAL it is not vulnerable because version (1.1.1.1.o-1) not 'less than' '1.1.1.j-1'. Then the NVD says it is vulnerable - `1.1.1.1.o-1` 'less than' **`1.1.1j`**. Similarly, as it has been discarded by OVAL, then the vulnerability is discarded: _not vulnerable to 'CVE-2021-23840' since it is not affected (feed 'OVAL')._
  3. [CVE-2022-1292](https://security.archlinux.org/CVE-2022-1292): OVAL considers it to be in `Unknown` status, so it will not be scanned for the vulnerability by OVAL. Then the NVD says it is vulnerable - `1.1.1.1.o-1` 'less than' **`1.1.1o`**, but as it has not been scanned by OVAL, it is discarded: _not vulnerable to 'CVE-2022-1292' since it is not affected (feed 'OVAL')._
  <br>

  - **`avahi` package** (different schema or lack of version information in the NVD):
  ```
  DEBUG: (5458): Package 'avahi' inserted into the vulnerability 'CVE-2021-3468'. Version (0.8+22+gfd482a7-3) 'greater than or equal' '0.8+22+gfd482a7-3' (feed 'OVAL').
  ...
  DEBUG: (5458): Package 'avahi' inserted into the vulnerability 'CVE-2021-26720'. Version (0.8+22+gfd482a7-3) 'less than or equal' '0.8-4' (feed 'NVD').
  DEBUG: (5459): Trying to insert duplicated package 'avahi' into the vulnerability 'CVE-2021-3468'. Version (0.8+22+gfd482a7-3) 'less than or equal' '0.8' (feed 'NVD').
  ...
  DEBUG: (5462): Package 'avahi' not vulnerable to 'CVE-2021-26720' since it is not affected (feed 'OVAL').
  ...
  DEBUG: (5468): The 'avahi' package (0.8+22+gfd482a7-3) from agent '001' is vulnerable to 'CVE-2021-3468'. Condition: 'Package greater than or equal 0.8+22+gfd482a7-3'
  ```
  1. [CVE-2021-3468](https://security.archlinux.org/CVE-2021-3468): From OVAL's side, it is vulnerable because version `0.8+22+gfd482a7-3` 'greater than or equal' `0.8+22+gfd482a7-3`. Then, the NVD also says that it is vulnerable, although with a bad match due to lack of information- `0.8+22+gfd482a7-3` 'less than or equal' `0.8`, although it can be seen that as the vulnerability already exists on the part of OVAL, then it would only reaffirm the vulnerability: _The 'avahi' package (0.8+22+gfd482a7-3) from agent '001' is vulnerable to 'CVE-2021-3468'._

  2. [CVE-2021-26720](https://security.archlinux.org/CVE-2021-26720): The vulnerability does not exist in OVAL, and the NVD says it is vulnerable by the following match (although it fails by the symbol) - `0.8+22+gfd482a7-3` 'less than or equal to' `0.8-4`. As it does not exist in the OVAL, then the vulnerability is discarded: _not vulnerable to 'CVE-2021-26720' since it is not affected (feed 'OVAL')._

**Comment by MarcelKemp** [comment]
## Conclusions

The vulnerability detector is **working correctly** in Arch Linux agent scanning, because in the cases discussed it responds as expected to avoid false positives.

- Vulnerabilities with `Unknown` status in the OVAL.

  It is normal that these vulnerabilities are not scanned, so that we avoid possible false positives, as they may be under analysis and may modify the versions that fix or affect the vulnerability. Currently, there are only 3 cases of vulnerabilities with unknown status:
  - https://security.archlinux.org/AVG-2725/log (created at 18 May 2022 20:00:07)
  - https://security.archlinux.org/AVG-2736/log (edited at 24 May 2022 06:40:34)
  - https://security.archlinux.org/AVG-2702/log (edited at 09 May 2022 23:08:41)

  Among those, the last case is under analysis in the NVD and was changed from Fixed to Unknown the last time it was modified.

  An example of a vulnerability with `Unknown` status in the past and currently in `Fixed` status: 
  https://security.archlinux.org/AVG-2759/log

  So, we hope that these vulnerabilities will change their _`status`_ as soon as possible.

  - To avoid confusion about such vulnerabilities, it is possible to add a new debug message informing that the vulnerability is not being scanned because it is in an `Unknown` state. 

- OVAL vulnerabilities with `Vulnerable` status and with a fixed version.

  It is normal to be considered `Vulnerable`, as the repository does not have the fixed version in production yet. 
  As soon as the new version is released, the vulnerability goes from `Vulnerable` to `Fixed` status.

- Problems with some versioning of NVD packages, according to the official OVAL.

  The OVAL is fully trusted for _Pacman_ packages, so it does not affect NVD versioning, thus avoiding false positives.
